### PR TITLE
Markdown: Preserve multiple spaces in inline code

### PR DESCRIPTION
**Pull request by kachkaev**
## Description

Fixes https://github.com/prettier/prettier/issues/13485

## Checklist

- [x] I’ve added tests to confirm my change works.
- ~~(If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).~~
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://deploy-preview-13590--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEADABAMwhdAjAQwCd8CAvdVEAGhAgAcYBLaAZ2VGKIgHcAFYgnYoCAGx4EAnu1p4iBMAGs4MAMoEAtnAAyTKHGSYxrOLPlKVq+gr0BzZDCIBXUyDga8cACZfv2glC2TgS2cABiEEQaBDDMgcggBE4wEDQgABYwGqIA6ulM8KzWYHCqQgVMAG4FkglgrDIgeiZEMHzyttGGxq4AVqwAHqp2onAAik4Q8N2iJrTWRC0J0USKXrxQafREejA5TF4w6cgAHAAM89wmOfL0CdtwLZUGtACOk-DtDMKJrAC0+m83jSRDg7yYoPaIS6SCMs1cJg0TAczgRI3GHwMsJ6tBgBDw+0OxyQACZcfImKI7ABhCAaGGJUSiNJOEwAFXxwjhcxAlRcAEkoL5YKowDtGABBIWqGCSUYzEwAX0VQA)✨**

### CODE REVIEW

File: src/language-markdown/print-preprocess.js [hunk 1]
```diff
@@ -34,7 +34,13 @@ function transformInlineCode(ast) {
       return node;
     }
 
-    return { ...node, value: node.value.replace(/\s+/g, " ") };
+    return {
+      ...node,
+      value: node.value
+        .replace(/^\s+/, " ")
+        .replace(/\n/g, " ")
+        .replace(/\s+$/, " "),
```
Comment (fisker) [inline src/language-markdown/print-preprocess.js:42]: Does `/(?:^\s+|\n|\s+$)/g` work? Is it faster?

Comment (kachkaev) [reply src/language-markdown/print-preprocess.js]: I‘ve updated the logic a bit. Looks like we need to treat ` ` and other whitespace differently, [according to the spec](https://spec.commonmark.org/0.30/#backtick-string).

File: tests/format/markdown/spec/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -1,5 +1,162 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`commonmark-0.30-example-328.md - {"proseWrap":"always"} format 1`] = `
```
Comment (kachkaev) [inline tests/format/markdown/spec/__snapshots__/jsfmt.spec.js.snap:3]: Context: https://github.com/prettier/prettier/issues/13591

**Comment by kachkaev** [comment]
Hmmm [Example 331](https://spec.commonmark.org/0.30/#example-331) fails second pass:

```
FULL_TEST=true yarn test markdown/spec/jsfmt
```

````txt
[space][space]``[space][space]
↓
[space]``[space]
↓
``
````

Screenshot 2022-10-08 at 21 37 09


Going to take a break now. Any thoughts?

**Comment by fisker** [comment]
example 331 & example 330 has different value, https://github.com/prettier/prettier/blob/62d95b33b515f83187aba245b3d97b937f37914d/src/language-markdown/clean.js#L36

**Comment by kachkaev** [comment]
Example 331 becomes Example 330 after first pass, does not it? Maybe I’m missing something.

File: src/language-markdown/print-preprocess.js [hunk 3]
```diff
@@ -31,7 +31,17 @@ function transformInlineCode(ast) {
       return node;
     }
 
-    return { ...node, value: node.value.replace(/\s+/g, " ") };
+    const valueWithoutNewlines = node.value.replace(/\n/g, " ");
```
Comment (thorn0) [inline src/language-markdown/print-preprocess.js:34]: Let's preserve new lines when `proseWrap` is set to `preserve'. This PR will supersede #11373 then.

Comment (thorn0) [reply src/language-markdown/print-preprocess.js:34]: #11373 has been merged. Please rebase.

**Comment by kachkaev** [comment]
Pressed the wrong button 😅

I included https://github.com/prettier/prettier/pull/11373. I still don’t fully understand the relationship between `src/language-markdown/print-preprocess.js` → `transformInlineCode` and `src/language-markdown/clean.js` 🤔 The latter does not have access to `options.proseWrap`, but tests still pass. If I remove any transforms from either function, some snapshots fail 🤔

**Comment by kachkaev** [comment]
@thorn0 I merged `next` into my branch, so PR playground includes https://github.com/prettier/prettier/pull/13606 now. Thanks for implementing ‘show preprocessed AST‘!

Do you foresee any other changes we need to make before merging?

File: src/language-markdown/print-preprocess.js [hunk 4]
```diff
@@ -27,11 +27,22 @@ function transformImportExport(ast) {
...
-    return { ...node, value: node.value.replace(/\s+/g, " ") };
+    const value =
+      options.proseWrap === "preserve"
+        ? node.value
+        : node.value.replace(/\n/g, " ");
+
+    return {
+      ...node,
+      value:
+        value.startsWith(" ") && value.endsWith(" ") && value.trim().length > 0
```
[the remark this answers is not in the document]
Comment (thorn0) [reply src/language-markdown/print-preprocess.js:42]: Okay, I see why you did this. But this should be done in `print`, not in `preprocess`.

[the remark this answers is not in the document]
Comment (kachkaev) [reply src/language-markdown/print-preprocess.js]: I see, thanks for the fix! 🙌

Do you know why we still need `newObj.value = ast.value.replace(/\n/g, " ");` in `clean.js`? This transform is applied regardless of `proseWrap`. I vaguely remember trying to remove it but tests failed.

File: changelog_unreleased/markdown/13590.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#### Preserve multiple spaces in inline code (#13590 by @kachkaev)
```
Comment (kachkaev) [inline changelog_unreleased/markdown/13590.md:1]:
```suggestion
#### Preserve multiple spaces in inline code (#13590 by @kachkaev and @thorn0)
```

@thorn0 feel free to commit this if you want

File: src/language-markdown/print-preprocess.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (thorn0) [reply src/language-markdown/print-preprocess.js]: Those tests are run only with `proseWrap: "always"`, that's why it works. Need to think what to do.

[the remark this answers is not in the document]
Comment (thorn0) [reply src/language-markdown/print-preprocess.js]: No, I was wrong. It's some other tests that fail if I add
```js
run_spec(import.meta, ["markdown"], { proseWrap: "preserve" });
run_spec(import.meta, ["markdown"], { proseWrap: "never" });
```
to `jsfmt.spec.js`, not related to code spans. 

As for `clean.js`, see how the `clean` function (AKA `massageAST`) is used:

https://github.com/prettier/prettier/blob/955553bb3972eae532206ffc3a7d4a2383cb7c62/src/cli/format.js#L137-L169

It's applied to both the original AST and to the AST built from Prettier's output, and then the ASTs are checked for deep equality. So it makes sense that this replacement is needed. After it, not matter what `proseWrap` is, the values in the two ASTs are equal.

**Review by thorn0 — APPROVED** [review]

File: src/language-markdown/print-preprocess.js [hunk 6]
```diff
@@ -31,7 +31,7 @@ function transformInlineCode(ast, options) {
       return node;
     }
 
-    return { ...node, value: node.value.replace(/\s+/g, " ") };
+    return { ...node, value: node.value.replace(/\n/g, " ") };
```
Comment (fisker) [inline src/language-markdown/print-preprocess.js:34]: Maybe we can move this to `print`? So we can reduce one `mapAst` call. Can be separate PR.

File: tests/format/js/multiparser-markdown/__snapshots__/jsfmt.spec.js.snap [hunk 7]
```diff
@@ -109,7 +109,7 @@ markdown\`
 
 =====================================output=====================================
 markdown\`
-  const cssString = css\\\` background-color: \\$\\{color('base')\\}\\\`;
+  const cssString = css\\\`  background-color: \\$\\{color('base')\\}\\\`;
```
Comment (fisker) [inline tests/format/js/multiparser-markdown/__snapshots__/jsfmt.spec.js.snap:112]: Shouldn't this call `markdown -> js -> css` and remove the leading space? Because it's not valid css?

**Review by fisker — APPROVED** [review]
Good job! Just a few unimportant questions.

File: tests/format/js/multiparser-markdown/__snapshots__/jsfmt.spec.js.snap (hunk 7 above)
Comment (thorn0) [reply tests/format/js/multiparser-markdown/__snapshots__/jsfmt.spec.js.snap:112]: This is a crazy test. `const cssString = ...` is Markdown here, not JS. The Markdown-in-JS embedder also includes logic for ignoring indentation (no idea what's the story behind that). That's why the output is like that.

Comment (fisker) [reply tests/format/js/multiparser-markdown/__snapshots__/jsfmt.spec.js.snap:112]: Ah, you explained before... forgot. Sorry.

**Comment by kachkaev** [comment]
Great, thanks for helping with this PR @thorn0 and @fisker! 👏
